[p2p] Pimpl AddrMan to abstract implementation details

amitiuttarwar opened the pull request:
Introduce the pimpl pattern for AddrMan to separate the implementation details from the externally used object representation. This reduces compile-time dependencies and conceptually clarifies AddrMan's interface from the implementation specifics. 

Since the unit & fuzz tests currently rely on accessing AddrMan internals, this PR introduces addrman_impl.h, which is exclusively imported by addrman.cpp and test files.

amitiuttarwar committed:
[net, addrman] Remove external dependencies on CAddrInfo objects

CAddrInfo objects are an implementation detail of how AddrMan manages and adds
metadata to different records. Encapsulate this logic by updating Select &
SelectTriedCollision to return the additional info that the callers need.

amitiuttarwar committed:
[addrman] Introduce CAddrMan::Impl to encapsulate addrman implementation.

Introduce the pimpl pattern for CAddrMan to separate the implementation details
from the externally used object representation. This reduces compile-time
dependencies and conceptually clarifies AddrMan's interface from the
implementation specifics.

Since the unit & fuzz tests currently rely on accessing CAddrMan internals, this
commit introduces addrman_impl.h, which is exclusively imported by addrman.cpp
and test files.

Review hint: git diff --color-moved=dimmed-zebra
--color-moved-ws=ignore-all-space

amitiuttarwar committed:
[move-only] Move constants to test-only header

Review hint: git diff --color-moved=dimmed-zebra
--color-moved-ws=ignore-all-space

amitiuttarwar committed:
[move-only] Move CAddrInfo to test-only header file

Now that no bitcoind callers require knowledge of the CAddrInfo object, it can
be moved into the test-only header file.

Review hint: use git diff --color-moved=dimmed-zebra
--color-moved-ws=ignore-all-space

amitiuttarwar committed:
[doc] Update comments

Maintain comments on the external interfaces rather than on the internal
functions that implement them.

amitiuttarwar committed:
[addrman] Change CAddrInfo access

Since knowledge of CAddrInfo is limited to callsites that import
addrman_impl.h, only objects in addrman.cpp or the tests have access. Thus we
can remove calling them friends and make the members public.

amitiuttarwar committed:
[includes] Fix up included files

amitiuttarwar committed:
scripted-diff: Rename CAddrMan to AddrMan

-BEGIN VERIFY SCRIPT-
git grep -l CAddrMan src/ test/ | xargs sed -i 's/CAddrMan/AddrMan/g'
-END VERIFY SCRIPT-

amitiuttarwar committed:
scripted-diff: Rename CAddrInfo to AddrInfo

-BEGIN VERIFY SCRIPT-
git grep -l CAddrInfo src/ | xargs sed -i 's/CAddrInfo/AddrInfo/g'
-END VERIFY SCRIPT-

amitiuttarwar committed:
[style] Run changed files through clang formatter.

naumenkogs commented on src/addrman.h:
```diff
@@ -99,7 +99,17 @@ class CAddrMan
+     *  decisions and gossiped to peers. Callers should be careful that updating
+     *  this information doesn't leak topology information to network spies.
+     *
+     *  net_processing calls this function when it *disconnects* from a peer to
```
How does this work exactly? Is it documented anywhere?

I assume you're talking about this code:
```
    if (node.fSuccessfullyConnected && misbehavior == 0 &&
        !node.IsBlockOnlyConn() && !node.IsInboundConn()) {
        // Only change visible addrman state for full outbound peers.  We don't
        // call Connected() for feeler connections since they don't have
        // fSuccessfullyConnected set.
        m_addrman.Connected(node.addr);
    }
```

amitiuttarwar commented:
Fixed the [lint failure](https://cirrus-ci.com/task/5301111516037120) that was caused by the circular dependencies check with 95a066efff8c63b7d81d93977c1980e019708fa9. The check treated `file_name.h` and `file_name.cpp` as the same module, so the import pattern here of `addrman.h` <- `addrman_impl.h` <- `addrman.cpp` was perceived as a circular dependency. I fixed this by telling it to also treat `file_name_impl.h` as part of the same module.

jnewbery commented:
Concept ACK. I'm a big fan of compilation firewall patterns. From Herb Sutter:

> One big advantage of this idiom is that it breaks compile-time dependencies. First, system builds run faster because using a Pimpl can eliminate extra #includes. I have worked on projects where converting just a few widely-visible classes to use Pimpls has halved the system’s build time. Second, it localizes the build impact of code changes because the parts of a class that reside in the Pimpl can be freely changed – that is, members can be freely added or removed – without recompiling client code. Because it’s so good at eliminating compilation cascades due to changes in only the now-hidden members, it’s often dubbed a “compilation firewall.”

(https://herbsutter.com/gotw/_100/)

Both are relevant here. addrman.h is eventually included by just about everything, so minimizing its compilation time is potentially a big win. There's also a plan to rework its implementation. If we first make addrman into a pimpl, that could theoretically be done without any impact on the rest of the codebase.

jonatack commented:
Concept ACK on improved separation and shorter build times. Just read through pages 147-156 of Scott Meyers' "Effective Modern C++" about the (unique) pointer to implementation idiom and started looking at the commit organisation, particularly the choices in the second ("Introduce CAddrMan::Impl to encapsulate addrman implementation") and third ("Remove external dependencies on CAddrInfo objects") commits where the action appears to be.

laanwj commented:
Concept ACK

> Both are relevant here. addrman.h is eventually included by just about everything, so minimizing its compilation time is potentially a big win.

Yes, at some point there was discussion about doing this for `txmempool.h` too, as it's included a fair bit and includes all these boost `multi_index` structures, which i'm sure slow down compilation quite a bit.

amitiuttarwar commented:
adopted a different fix to the detected circular dependency- I made the impl a stand-alone class, so it's now `AddrManImpl` instead of `AddrMan::Impl`. this means `addrman_impl.h` does not need to import `addrman.h`, and the linter script does not need to be updated :)

JeremyRubin commented:
concept ack!

amitiuttarwar commented:
thanks for the concept ACKs! rebased to incorporate the recent addrman changes.

amitiuttarwar commented:
I had missed a silent merge conflict, fixed now and ready for review 🎈

jnewbery commented on src/addrman.cpp:
```diff
@@ -1004,16 +984,187 @@ CAddrInfo CAddrMan::SelectTriedCollision_()
+template void AddrMan::Unserialize(CHashVerifier<CDataStream>& s);
+
+template <typename Stream>
+void AddrMan::Serialize(Stream& s_) const
```
This may just be personal preference, but I think this file would be tidier/easier to read if all of these `AddrMan` boilerplate forwarding functions were collected at the bottom (as is done in txrequest.cpp), rather than interleaved with `AddrManImpl` functions.

amitiuttarwar replied:
I agree, updated

jnewbery commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,287 @@
+    int ForceCheckAddrman() const EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    /**
+     * This internal function adds one parameter to the GetAddr wrapper.
+     *
+     * @param[out] vAddr         Vector of randomly selected addresses from vRandom.
+     */
```
Maybe we should just update the internal `GetAddr_()` function to return a `std::vector<CAddress>` instead of introducing this comment?

amitiuttarwar replied:
sounds good, added commit 5a8415750cdc93b719a8e4c18ee245e236b93006 to handle that

jnewbery commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,287 @@
+
+#include <logging.h>
+
+#include <sync.h>
```
Should be grouped above with logging.h, since sync.h is a project header and not a stl header.

amitiuttarwar replied:
done

jnewbery commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,287 @@
+
+#include <sync.h>
+#include <unordered_map>
+#include <unordered_set>
```
Also include:

- cstdint (for fixed width integer types)
- netaddress.h (for CNetAddr)
- serialize.h (for serialization functions)
- protocol.h (for CAddress)
- uint256.h (for uint256)
- vector (for std::vector)
- utility (for std::pair)
- set (for std::set)
- optional (for std::optional)

amitiuttarwar replied:
done

jnewbery commented on src/addrman.h:
```diff
@@ -6,94 +6,19 @@
 #include <netaddress.h>
 #include <protocol.h>
-#include <sync.h>
 #include <timedata.h>
```
Also include streams.h

amitiuttarwar replied:
done

jnewbery reviewed (commented):
Code review ACK 11926b09935f491c33b611f1a112a800377a481a. Just a few very minor comments inline, and a few commit order comments:

1.

In the first commit (_[addrman] Introduce CAddrMan::Impl to encapsulate addrman implementation._), you slightly change the function bodies of `SelectTriedCollision()` and `Select()`:

```diff
         Check();
-        const CAddrInfo ret = SelectTriedCollision_();
+        const auto ret = SelectTriedCollision_();
         Check();
```

and

```diff
         Check();
-        const CAddrInfo addrRet = Select_(newOnly);
+        const auto addrRet = Select_(newOnly);
         Check();
```

at the same time as moving that code from addrman.h to addrman.cpp. That means that it doesn't show up as moved code when using `git diff --color-moved=dimmed-zebra`. I think those changes to `auto` can happen in the second commit (_[net, addrman] Remove external dependencies on CAddrInfo objects_)

2.

You add `#include <addrman_impl.h>` to fuzz/deserialize.cpp in the first commit (_[addrman] Introduce CAddrMan::Impl to encapsulate addrman implementation._). I think it's only needed there from the 3rd commit (_[move-only] Move CAddrInfo to test-only header file_), since the include is only needed for the definition of `CAddrInfo`.

3.

The updates to the `Select()` and `SelectTriedCollision()` doxygen comments should be included in the changes to the interface in commit 2 (_[net, addrman] Remove external dependencies on CAddrInfo objects)_

4.

In commit _scripted-diff: Rename CAddrMan to AddrMan_, is there a reason not to use `git grep -l CAddrMan`?

amitiuttarwar committed:
[refactor] Update GetAddr_() function signature

Update so the internal function signature matches the external one, as is the
case for the other addrman functions.

amitiuttarwar commented on src/addrman.h:
```diff
@@ -123,290 +48,63 @@ class CAddrInfo : public CAddress
- * Stochastical (IP) address manager
- */
-class CAddrMan
+class AddrManImpl;
```
fixed

amitiuttarwar commented:
thanks for the review @jnewbery. I took all your suggestions except for this one that I have a question about-

> In commit scripted-diff: Rename CAddrMan to AddrMan, is there a reason not to use git grep -l CAddrMan?

I was trying this out, but am unsure how to exclude release notes. This is what I have so far, any suggestions? `git grep -l CAddrMan | xargs sed -i 's/CAddrMan/AddrMan/g'`

also, thank you for the feedback about commit breakdown. I hope its all straightened out now :)

jamesob commented:
If this is being done in the name of compilation performance, do you have any measurements justifying the change?

MarcoFalke commented:
> I was trying this out, but am unsure how to exclude release notes. 

`git grep -l "foo_bar" ./src`

jnewbery commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,291 @@
+class AddrManImpl
+{
+public:
+    AddrManImpl(std::vector<bool>&& asmap, bool deterministic, int32_t consistency_check_ratio)
```
What do you think about having the ctor and dtor function bodies in addrman.cpp, so that addrman_impl.h only contains the function declarations?

amitiuttarwar replied:
done

jnewbery commented on src/addrman.cpp:
```diff
@@ -1004,16 +987,184 @@ CAddrInfo CAddrMan::SelectTriedCollision_()
+    return m_asmap;
+}
+
+// explicit instantiation
```
It seems more natural to have these instantiations below the function definitions for `Serialize()` and `Unserialize()`.

amitiuttarwar replied:
to make sure I understand correctly, is this the order you're suggesting? 
- `AddrManImpl::Serialize`
- `AddrManImpl::Unserialize`
- explicit instantiations
- `AddrManImpl` [ other functions ]
- `AddrMan::Serialize`
- `AddrMan::Unserialize`

The reason I think the current order makes more sense is because the explicit instantiations are for the `AddrMan` template, not the impl template.

jnewbery replied:
I mean that these instantiations of `AddrMan::Serialize()` and `AddrMan::Unserialize()` should be below the template function definitions for `AddrMan::Serialize()` and `AddrMan::Unserialize()` (i.e. just move them down 10 lines).

amitiuttarwar replied:
ahhhh, that makes more sense. done

jnewbery commented on src/addrman.cpp:
```diff
@@ -101,26 +103,13 @@ double CAddrInfo::GetChance(int64_t nNow) const
-        }
-    }
-}
+AddrMan::AddrMan(std::vector<bool> asmap, bool deterministic, int32_t consistency_check_ratio)
```
Could these `AddrMan` ctor and dtor methods also be moved to the bottom, below the `AddrManImpl` methods?

amitiuttarwar replied:
done

jnewbery commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,291 @@
+
+    std::pair<CAddress, int64_t> SelectTriedCollision_() EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    //! Consistency check, taking into account m_consistency_check_ratio. Will std::abort if an inconsistency is detected.
+    void Check() const EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    //! Perform consistency check, regardless of m_consistency_check_ratio.
+    //! @returns an error code or zero.
+    int ForceCheckAddrman() const EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Perhaps these two methods should be moved above `Good_`, so that all the underscore internal functions are grouped together.

amitiuttarwar replied:
hm, but they are still internal functions?

jnewbery replied:
By "internal functions", I mean private functions that are called by the public "outer functions". There's a pattern that the public `Good()` function calls private `Good_()`, public `Add()` calls private `Add_()`, etc. It makes sense to group all of the `FunctionName_()` functions together.

amitiuttarwar replied:
okay, I introduced this new commit: https://github.com/bitcoin/bitcoin/pull/22950. I moved `Check` and `ForceCheck` to be after all the `Function_`. And I moved some additional stuff around to be consistently ordered in declarations & definitions.

jnewbery reviewed (commented):
A couple more suggestions for the git ordering/commit logs:

1.

I think the first commit _[addrman] Introduce CAddrMan::Impl to encapsulate addrman implementation._ could be split into two commits, to make it even easier for reviewers to verify that there have been no changes. The first commit would move all of the function bodies (eg the outer `size()` ,`Add()`, `Good()`, etc) out of addrman.h into addrman.cpp and could be very easily reviewed with `git diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space`. The second commit would add the pimpl. That means the commit isn't a mixture of new code, renames and moves.

2.

The commit log for _[doc] Update comments_ includes "Remove them from the internal ones unless there is a divergence in function behaviors.", which can now be removed, since there are no longer any instances where the behaviour diverges.

jnewbery commented:
> If this is being done in the name of compilation performance, do you have any measurements justifying the change?

@jamesob - the most significant impact would be on _recompilation_ times. By totally separating the implementation from the interface, we can make changes to the implementation without any impact on client code that calls into this component. For example, on this branch, making the following change:

<details>
<summary>Code change</summary>

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index ac2635249c..1257a7503f 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -1015,7 +1015,7 @@ bool AddrManImpl::Add(const std::vector<CAddress>& vAddr, const CNetAddr& source
         nAdd += Add_(*it, source, nTimePenalty) ? 1 : 0;
     Check();
     if (nAdd) {
-        LogPrint(BCLog::ADDRMAN, "Added %i addresses from %s: %i tried, %i new\n", nAdd, source.ToString(), nTried, nNew);
+        LogPrint(BCLog::ADDRMAN, "Added %i addresses from %s: %i tried, %i new, %i collisions\n", nAdd, source.ToString(), nTried, nNew, m_tried_collisions.size());
     }
     return nAdd > 0;
 }
```
</details>

Recompiles for me in about 35s:

<details>
<summary>Build</summary>

```
→ time make
Making all in src
make[1]: Entering directory '/home/vagrant/bitcoin/src'
make[2]: Entering directory '/home/vagrant/bitcoin/src'
make[3]: Entering directory '/home/vagrant/bitcoin'
make[3]: Leaving directory '/home/vagrant/bitcoin'
  CXX      libbitcoin_server_a-addrman.o
  AR       libbitcoin_server.a
  CXXLD    bitcoind
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  CXXLD    test/fuzz/fuzz
make[2]: Leaving directory '/home/vagrant/bitcoin/src'
make[1]: Leaving directory '/home/vagrant/bitcoin/src'
Making all in doc/man
make[1]: Entering directory '/home/vagrant/bitcoin/doc/man'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/vagrant/bitcoin/doc/man'
make[1]: Entering directory '/home/vagrant/bitcoin'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/vagrant/bitcoin'

real    0m35.758s
user    0m26.836s
sys     0m7.951s
```

</details>

On master, making the equivalent change:

<details>
<summary>Code change</summary>

```diff
diff --git a/src/addrman.h b/src/addrman.h
index 7dd8528bef..08f158f6c6 100644
--- a/src/addrman.h
+++ b/src/addrman.h
@@ -174,7 +174,7 @@ public:
             nAdd += Add_(*it, source, nTimePenalty) ? 1 : 0;
         Check();
         if (nAdd) {
-            LogPrint(BCLog::ADDRMAN, "Added %i addresses from %s: %i tried, %i new\n", nAdd, source.ToString(), nTried, nNew);
+            LogPrint(BCLog::ADDRMAN, "Added %i addresses from %s: %i tried, %i new, %i collisions\n", nAdd, source.ToString(), nTried, nNew, m_tried_collisions.size());
         }
         return nAdd > 0;
     }
```

</details>

results in a very long build of 9m29s, since those changes get included in many other translation units:

<details>
<summary>Build</summary>

```
→ time make
Making all in src
make[1]: Entering directory '/home/vagrant/bitcoin/src'
make[2]: Entering directory '/home/vagrant/bitcoin/src'
make[3]: Entering directory '/home/vagrant/bitcoin'
make[3]: Leaving directory '/home/vagrant/bitcoin'
  CXX      init/bitcoind-bitcoind.o
  CXX      libbitcoin_server_a-addrdb.o
  CXX      libbitcoin_server_a-addrman.o
  CXX      libbitcoin_server_a-init.o
  CXX      libbitcoin_server_a-mapport.o
  CXX      libbitcoin_server_a-net.o
  CXX      libbitcoin_server_a-net_processing.o
  CXX      node/libbitcoin_server_a-context.o
  CXX      node/libbitcoin_server_a-interfaces.o
  CXX      node/libbitcoin_server_a-transaction.o
  CXX      rpc/libbitcoin_server_a-mining.o
  CXX      rpc/libbitcoin_server_a-net.o
  CXX      libbitcoin_server_a-torcontrol.o
  CXX      libbitcoin_server_a-txorphanage.o
  CXX      libbitcoin_server_a-txrequest.o
  CXX      wallet/libbitcoin_server_a-init.o
  AR       libbitcoin_server.a
  CXX      interfaces/libbitcoin_util_a-init.o
  AR       libbitcoin_util.a
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-wallet
  CXXLD    bitcoin-util
  CXX      test/test_bitcoin-addrman_tests.o
  CXX      test/test_bitcoin-denialofservice_tests.o
  CXX      test/test_bitcoin-i2p_tests.o
  CXX      test/test_bitcoin-validation_tests.o
  CXX      test/test_bitcoin-net_peer_eviction_tests.o
  CXX      test/test_bitcoin-net_tests.o
  CXX      test/test_bitcoin-txrequest_tests.o
  CXX      test/util/libtest_util_a-net.o
  CXX      test/util/libtest_util_a-setup_common.o
  AR       libtest_util.a
  CXXLD    test/test_bitcoin
  CXX      test/fuzz/fuzz-addition_overflow.o
  CXX      test/fuzz/fuzz-addrman.o
  CXX      test/fuzz/fuzz-autofile.o
  CXX      test/fuzz/fuzz-banman.o
  CXX      test/fuzz/fuzz-block_header.o
  CXX      test/fuzz/fuzz-blockfilter.o
  CXX      test/fuzz/fuzz-bloom_filter.o
  CXX      test/fuzz/fuzz-buffered_file.o
  CXX      test/fuzz/fuzz-chain.o
  CXX      test/fuzz/fuzz-checkqueue.o
  CXX      test/fuzz/fuzz-coins_view.o
  CXX      test/fuzz/fuzz-connman.o
  CXX      test/fuzz/fuzz-crypto.o
  CXX      test/fuzz/fuzz-crypto_aes256.o
  CXX      test/fuzz/fuzz-crypto_aes256cbc.o
  CXX      test/fuzz/fuzz-crypto_chacha20.o
  CXX      test/fuzz/fuzz-crypto_chacha20_poly1305_aead.o
  CXX      test/fuzz/fuzz-crypto_common.o
  CXX      test/fuzz/fuzz-crypto_hkdf_hmac_sha256_l32.o
  CXX      test/fuzz/fuzz-crypto_poly1305.o
  CXX      test/fuzz/fuzz-cuckoocache.o
  CXX      test/fuzz/fuzz-deserialize.o
  CXX      test/fuzz/fuzz-fee_rate.o
  CXX      test/fuzz/fuzz-fees.o
  CXX      test/fuzz/fuzz-flatfile.o
  CXX      test/fuzz/fuzz-float.o
test/fuzz/float.cpp: In function ‘void float_fuzz_target(FuzzBufferType)’:
test/fuzz/float.cpp:60:30: warning: ‘tmp’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   60 |         assert(std::isnan(d) || d == d_deserialized);
      |                              ^~
  CXX      test/fuzz/fuzz-golomb_rice.o
  CXX      test/fuzz/fuzz-http_request.o
  CXX      test/fuzz/fuzz-i2p.o
  CXX      test/fuzz/fuzz-integer.o
  CXX      test/fuzz/fuzz-kitchen_sink.o
  CXX      test/fuzz/fuzz-load_external_block_file.o
  CXX      test/fuzz/fuzz-merkleblock.o
  CXX      test/fuzz/fuzz-message.o
  CXX      test/fuzz/fuzz-muhash.o
  CXX      test/fuzz/fuzz-multiplication_overflow.o
  CXX      test/fuzz/fuzz-net.o
  CXX      test/fuzz/fuzz-net_permissions.o
  CXX      test/fuzz/fuzz-netaddress.o
  CXX      test/fuzz/fuzz-netbase_dns_lookup.o
  CXX      test/fuzz/fuzz-node_eviction.o
  CXX      test/fuzz/fuzz-p2p_transport_serialization.o
  CXX      test/fuzz/fuzz-parse_hd_keypath.o
  CXX      test/fuzz/fuzz-policy_estimator.o
  CXX      test/fuzz/fuzz-policy_estimator_io.o
  CXX      test/fuzz/fuzz-pow.o
  CXX      test/fuzz/fuzz-primitives_transaction.o
  CXX      test/fuzz/fuzz-process_message.o
  CXX      test/fuzz/fuzz-process_messages.o
  CXX      test/fuzz/fuzz-protocol.o
  CXX      test/fuzz/fuzz-random.o
  CXX      test/fuzz/fuzz-rbf.o
  CXX      test/fuzz/fuzz-rolling_bloom_filter.o
  CXX      test/fuzz/fuzz-rpc.o
  CXX      test/fuzz/fuzz-script.o
  CXX      test/fuzz/fuzz-script_bitcoin_consensus.o
  CXX      test/fuzz/fuzz-script_descriptor_cache.o
  CXX      test/fuzz/fuzz-script_interpreter.o
  CXX      test/fuzz/fuzz-script_ops.o
  CXX      test/fuzz/fuzz-script_sigcache.o
  CXX      test/fuzz/fuzz-script_sign.o
  CXX      test/fuzz/fuzz-scriptnum_ops.o
  CXX      test/fuzz/fuzz-secp256k1_ec_seckey_import_export_der.o
  CXX      test/fuzz/fuzz-secp256k1_ecdsa_signature_parse_der_lax.o
  CXX      test/fuzz/fuzz-signature_checker.o
  CXX      test/fuzz/fuzz-signet.o
  CXX      test/fuzz/fuzz-socks5.o
  CXX      test/fuzz/fuzz-span.o
  CXX      test/fuzz/fuzz-string.o
  CXX      test/fuzz/fuzz-strprintf.o
  CXX      test/fuzz/fuzz-system.o
  CXX      test/fuzz/fuzz-timedata.o
  CXX      test/fuzz/fuzz-torcontrol.o
  CXX      test/fuzz/fuzz-tx_pool.o
  CXX      test/fuzz/fuzz-txrequest.o
  CXX      test/fuzz/fuzz-utxo_snapshot.o
  CXX      test/fuzz/fuzz-validation_load_mempool.o
  CXX      test/fuzz/fuzz-versionbits.o
  CXX      test/fuzz/libtest_fuzz_a-util.o
  AR       libtest_fuzz.a
  CXXLD    test/fuzz/fuzz
make[2]: Leaving directory '/home/vagrant/bitcoin/src'
make[1]: Leaving directory '/home/vagrant/bitcoin/src'
Making all in doc/man
make[1]: Entering directory '/home/vagrant/bitcoin/doc/man'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/vagrant/bitcoin/doc/man'
make[1]: Entering directory '/home/vagrant/bitcoin'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/vagrant/bitcoin'

real	9m29.276s
user	8m34.052s
sys	0m51.490s

```

</details>

(gui and bench were disabled for all builds - I expect the difference would be even more pronounced with those components also enabled).

There may be a marginal speed up in compiling from scratch since the header is slimmed down, but I expect it'd be a small improvement.

However, for me the improved recompilation time is **not** the main motivation for this change. There are two motivations which I think are more important:

1. Slimming down the header file to just well-documented interface methods makes it much easier for people writing client code that relies on addrman to very quickly understand the class's responsibilities and contracts without having to concern themselves with the internal implementation details.
2. The upcoming rework of addrman (https://github.com/sipa/bitcoin/tree/202106_multiindex_addrman) can be achieved with _just_ changes to addrman.cpp, leaving the interface and header file the same. That makes it much easier to review, verify that there are no functional changes, and test conformance between the old and new implementations.

amitiuttarwar committed:
[move-only] Move CAddrMan function definitions to cpp

In preparation for introducing the pimpl pattern to addrman, move all function
bodies out of the header file.

Review hint: use git diff --color-moved=dimmed-zebra
--color-moved-ws=ignore-all-space

amitiuttarwar commented:
latest push: 
- rebased to address a small conflict with #23025 
- updated the scripted diffs to use `grep` to generate the file names 
- split up the first commit into two parts as suggested in https://github.com/bitcoin/bitcoin/pull/22950#pullrequestreview-761829604
- moved around some code to address review comments

mzumsande reviewed (commented):
Concept ACK

I did a first round of review of the changes and it looks really good to me, just some minor comments below.

One general point: I wonder whether the pattern `AddrManImpl::Function()` and `AddrManImpl::Function_()` should still exist in the context of the PImpl, because now there are already two sets of functions that mostly just pass things along. With the recent changes to `Check()` allowing much more flexibility when and how often to call the checks, maybe it is not necessary anymore and the `Check()` calls could be merged into the `Function_()` ? (Changing this shouldn't be part of this PR, just wanted to bring up the idea).

a typo: "objecjt" in 527cbd168acb776eb03e4506f186afb5e0a85d90 commit message

martinus commented on src/addrman_impl.h:
```diff
@@ -0,0 +1,206 @@
+
+    size_t size() const EXCLUSIVE_LOCKS_REQUIRED(!cs);
+
+    bool Add(const std::vector<CAddress> &vAddr, const CNetAddr& source, int64_t nTimePenalty = 0)
```
nit: I'd remove all the default arguments in AddrManImpl, at least in the public interface. Having defaults in both `AddrMan` and `AddrManImpl` can be a source of error when not both are updated

amitiuttarwar replied:
definitely agree with this feedback! disagree with classifying it as a nit :)

I believe I removed all the defaults from the public interface, but I left them on the private functions because they aren't redundant in that situation.

jnewbery commented:
> One general point: I wonder whether the pattern AddrManImpl::Function() and AddrManImpl::Function_() should still exist in the context of the PImpl, because now there are already two sets of functions that mostly just pass things along. With the recent changes to Check() allowing much more flexibility when and how often to call the checks, maybe it is not necessary anymore and the Check() calls could be merged into the Function_() ? (Changing this shouldn't be part of this PR, just wanted to bring up the idea).

@mzumsande - I agree that this would be nice. Once this PR has been merged, that kind of change can be made without touching the header file and recompiling anything.

Do you think it would be overkill to have a variadic template `CheckExecuteCheck()` function that implements the pattern of locking `cs`, running `Check()`, calling the internal function, then running `Check()` again?

jnewbery commented:
utACK 0cc7031002079b35953bac1e9d4268b6321893c7

If you retouch again, you could move the function bodies of `CAddrMan::~CAddrMan()` and `CAddrMan::GetAsmap()` into addrman.cpp in the first commit (_[move-only] Move CAddrMan function definitions to cpp_) to make the second commit cleaner.

theuni commented on src/addrman.cpp:
```diff
@@ -1080,10 +1083,9 @@ std::vector<CAddress> AddrManImpl::GetAddr(size_t max_addresses, size_t max_pct,
     Check();
-    std::vector<CAddress> vAddr;
-    GetAddr_(vAddr, max_addresses, max_pct, network);
+    const auto addresses = GetAddr_(max_addresses, max_pct, network);
```
Nit: If the compiler doesn't elide this, the const will force an extra copy: https://stackoverflow.com/questions/25784544/will-returning-a-const-object-from-a-function-prevent-move-construction-from-out/25786015

Is there any specific reason for it?

amitiuttarwar replied:
that's so interesting, I hadn't thought about the correlation between copy elision and temporary const variables. thanks for digging in! 

next question is how I get my hands on the spec..

sipa replied:
It certainly looks like in practice the const doesn't prevent elision: https://godbolt.org/z/Go8xhGP1b

amitiuttarwar committed:
[move-only] Match ordering of CAddrMan declarations and definitions

Also move `Check` and `ForceCheckAddrman` to be after the `FunctionName_` functions.

Review hint: use git diff --color-moved=dimmed-zebra
--color-moved-ws=ignore-all-space

amitiuttarwar commented on src/addrman.cpp:
```diff
@@ -1010,9 +1010,8 @@ CAddrInfo AddrManImpl::SelectTriedCollision_()
-    int id_old = vvTried[tried_bucket][tried_bucket_pos];
-
-    return mapInfo[id_old];
+    auto info_old = mapInfo[vvTried[tried_bucket][tried_bucket_pos]];
```
makes sense, done

amitiuttarwar commented on src/addrman.cpp:
```diff
@@ -694,15 +694,13 @@ void AddrManImpl::Attempt_(const CService& addr, bool fCountFailure, int64_t nTi
 
-    if (vRandom.empty())
-        return CAddrInfo();
+    if (vRandom.empty()) return {};
```
great point, thanks for verifying & sharing!

amitiuttarwar commented on src/addrman_impl.h:
```diff
@@ -5,6 +5,77 @@
+{
+private:
+    //! last try whatsoever by us (memory only)
+    int64_t nLastTry{0};
```
fixed it

amitiuttarwar commented:
Thank you for the thoughtful reviews @mzumsande, @martinus, @jnewbery & @theuni !! I believe I've incorporated all the review comments. 

**A note on the latest push:** 

Mostly I rebased & addressed comments, but in doing so I added another commit- f2e5f38f09ee40933f752680fe7d75ee8e529fae. I generally attempted to get the grouping / ordering of functions to be consistent, between the declarations & definitions in AddrMan & AddrManImpl. Since (for now) the functions are repeated a few times, having predictable ordering seems valuable for making the file easier for human parsing.

**Thoughts on future work:** 

1. I agree it would be great to reduce the boilerplate code of having all interface functions 3 times. 

@jnewbery, @mzumsande, @theuni 
> With the recent changes to Check() allowing much more flexibility when and how often to call the checks, maybe it is not necessary anymore and the Check() calls could be merged into the Function_() ?

> Do you think it would be overkill to have a variadic template CheckExecuteCheck() function that implements the pattern of locking cs, running Check(), calling the internal function, then running Check() again?

Hehe, I've been meaning to play out the variadic template and take a look, it seems very appealing to reduce the boilerplate code. That said, I think incorporating `Check()` into `Function_()` would also be an improvement. 

2. Improving the interface of `Select` & `SelectTriedCollision`

@theuni 
> It would be nice if we had a time type so it'd be obvious what the second member of the pair was in the return values from Select and friends

I agree, it's unfortunate that this PR made things slightly worse because addrman's interface now exposes an `int64_t` to represent time. I opted for this solution over making it a chrono in an attempt to keep the changes of this PR more mechanical. 

Both of these functions are exclusively used by `CConnman::ThreadOpenConnections`, and I think the best solution would be to simplify the interfaces. When connman identifies that it is time to open a feeler, I think it should simply ask addrman for the top priority address. The concept of prioritizing tried collisions over a random connection should be internal to addrman. 

Making the code change is not difficult, but we'd have to make the change with care because it would logically change the handling of repeated failures [here](https://github.com/bitcoin/bitcoin/blob/3c776fdcec176ffaa2056633fa2b4e737cda29ce/src/net.cpp#L2049). So, this is a long winded way of saying, I hope returning a pair is a temporary thing.

jnewbery commented:
ACK 021f86953e8a1dff8ecc768186368d345c865cc2

The incremental commits make this really easy to review. Thank you!

rajarshimaitra commented on src/net.h:
```diff
@@ -797,7 +797,7 @@ class CConnman
     }
 
-    CConnman(uint64_t seed0, uint64_t seed1, CAddrMan& addrman, bool network_active = true);
+    CConnman(uint64_t seed0, uint64_t seed1, AddrMan& addrman, bool network_active = true);
```
As we are dropping the `C` from `AddrMan` and `AddrInfo`, is that a convention that will be applied elsewhere in the codebase from now on too? 

My understanding was the `C` before these structures indicated they are a `class`. Are we deciding to drop that indication from now on? 

Just trying to understand the rationale of name changes in https://github.com/bitcoin/bitcoin/pull/22950/commits/dd8f7f250095e58bbf4cd4effb481b52143bd1ed and https://github.com/bitcoin/bitcoin/pull/22950/commits/375750387e35ed751d1f5ab48860bdec93977f64

martinus replied:
Since https://github.com/bitcoin/bitcoin/pull/10461 the `developer-notes.md` states "Do not prefix class names with `C`"

michaelfolkson replied:
@rajarshimaitra: See https://bitcoin.stackexchange.com/questions/106558/what-do-the-classes-chainstatemanager-cchainstate-cchain-and-blockmanager-do-i or the Core PR review club that links to. New classes shouldn't have C in front of them but old ones aren't being renamed.

rajarshimaitra replied:
Oh ok.. Thanks for the clarification. 
Just to understand further, its ok for old ones to drop the `C` in later PRs if they are being touched right? Just like the case here with `AddrMan` and `AddrInfo`?

So eventually the goal is to remove "C for classes" everywhere?

sipa replied:
New code should follow the style described in the developer notes. If you're substantially changing a piece of code, it makes sense to also update it to follow that convention.

rajarshimaitra reviewed (commented):
Concept + Code Review ACK https://github.com/bitcoin/bitcoin/pull/22950/commits/021f86953e8a1dff8ecc768186368d345c865cc2

To check the effect of compilation firewall, I added a private integer member in `CAddrMan` of `master` and `AddrManImpl` of #22950. 

Verified that the addition causes recompilation of many tests and and other object files in `master`, where as in case of #22950, only the compilation of relevant test files (that includes addrman_impl header) are triggered..

Overall the commit changes are very clear and easy to review. Compilation time reduction is significant and very much observable.

Also thanks for such an elaborate review club description. Helped me to learn a ton about compiler firewall.

Below is one naming question for my own understanding.

GeneFerneau commented:
utACK [021f869](https://github.com/bitcoin/bitcoin/pull/22950/commits/021f86953e8a1dff8ecc768186368d345c865cc2)

mzumsande commented:
ACK 021f86953e8a1dff8ecc768186368d345c865cc2

Reviewed the code and did some sanity checks running this for a while on mainnet with `-checkaddrman=1`, but no testing beyond that.

theuni commented on src/addrman.cpp:
```diff
@@ -759,8 +759,9 @@ void AddrManImpl::GetAddr_(std::vector<CAddress>& vAddr, size_t max_addresses, s
 
     // gather a list of random nodes, skipping those of low quality
     const int64_t now{GetAdjustedTime()};
+    std::vector<CAddress> addresses;
```
Tiny nit for one of the refactor PRs: `addresses.reserve(std::min(vRandom.size(), nNodes));`

theuni reviewed (approved):
ACK 021f86953e8a1dff8ecc768186368d345c865cc2

I had a last-minute panic that some of the unsuspicious moves from CAddrInfo to CAddress could quietly change the behavior of calls like `addr.IsValid()` if `isValid` had previously been overridden by CAddrInfo, but fortunately no such overrides existed.

theStack reviewed (commented):
Post-merge concept and code-review ACK 021f86953e8a1dff8ecc768186368d345c865cc2
